core/src/main/java/io/questdb/network/IODispatcherOsx.java

```diff
@@ -75,14 +101,84 @@ private void enqueuePending(int watermark) {
         }
     }
 
-    private int findPending(long ts) {
-        return pending.binarySearch(ts, M_ID);
+    private boolean handleSocketOperation(long id) {
+        // find row in pending for two reasons:
+        // 1. find payload
+        // 2. remove row from pending, remaining rows will be timed out
+        final int row = pending.binarySearch(id, OPM_ID);
+        if (row < 0) {
+            LOG.error().$("internal error: kqueue returned unexpected id [id=").$(id).I$();
+            return false;
+        }
+
+        final C context = pending.get(row);
+        final SuspendEvent suspendEvent = context.getSuspendEvent();
+        if (suspendEvent != null) {
+            // the operation is suspended, check if we have a client disconnect
+            if (testConnection(context.getFd())) {
+                doDisconnect(context, id, DISCONNECT_SRC_PEER_DISCONNECT);
+                pending.deleteRow(row);
+                return true;
+            }
+        } else {
+            publishOperation(
+                    kqueue.getFilter() == KqueueAccessor.EVFILT_READ ? IOOperation.READ : IOOperation.WRITE,
+                    context
+            );
+            pending.deleteRow(row);
+            return true;
+        }
+        return false;
+    }
+
+    private void handleSuspendEvent(long id) {
+        final int eventsRow = pendingEvents.binarySearch(id, EVM_ID);
+        if (eventsRow < 0) {
+            LOG.error().$("internal error: kqueue returned unexpected event id [eventId=").$(id).I$();
+            return;
+        }
+
+        final long opId = pendingEvents.get(eventsRow, EVM_OPERATION_ID);
+        final int row = pending.binarySearch(opId, OPM_ID);
+        if (row < 0) {
+            LOG.error().$("internal error: suspended operation not found [id=").$(opId).$(", eventId=").$(id).I$();
```

Comment:
should be critical

Reply:
Changed in bbec37baf832db9a846d04e59ab124f8bb84eadc